tests/ui/iter_nth_zero.rs

```diff
@@ -29,3 +29,17 @@ fn main() {
     let mut iter2 = s3.iter();
     let _unwrapped = iter2.nth(0).unwrap();
 }
+
+struct Issue9820;
+
+impl Iterator for Issue9820 {
+    type Item = ();
+
+    fn nth(&mut self, _n: usize) -> Option<Self::Item> {
+        todo!()
+    }
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.nth(0)
```

Comment:
nit: include a comment saying that we shouldn't lint here (and a short blurb on why)